Add Chameleon PDA

SlamBamActionman opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->

This PR adds the Chameleon PDA, a PDA that has Chameleon capabilities. 

The Chameleon PDA is added to the Thief's Chameleon Kit and the filled Chameleon Backpack.

PDAs also no longer shows the name of the person it spawned on.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->

The Chameleon PDA is basically a necessity for any disguise to pass any amount of inspection, and allows for more advanced disguises for things such as ERT and CentCom. This will also allow StealthOps Nukies to much more convincingly blend in with crew instead of getting a Clown PDA or wearing the Agent ID without a PDA.

~~Since the Chameleon PDA does not come with an "Owner", to keep players from metagaming it that line in the UI has been removed. We already have IDs for identification, PDAs having a unique unmodifiable name set when looking at the UI is overkill.~~

PDA Owners are now overwritten whenever a new ID is inserted. This means you can still track who the owner is of a PDA if the ID is removed, however it protects Chameleon PDAs and can also enable some antag sabotage if they're willing to go that far. 

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

PdaSystem visuals have been moved over to PdaVisualizerSystem, and the state uses AppearanceData instead of a component property.

Chameleon clothes now also copies AppearanceData from the prototype it is copying.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

![image](https://github.com/user-attachments/assets/c4e0883c-442f-41dd-a67b-b0afa42e9806)

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

:cl:
- add: Added Chameleon PDA, found in the Thief & Uplink Chameleon Kit!
- tweak: The PDA "owner" field now updates upon having a new ID inserted.

BombasterDS commented:
Maybe it's reasonable decision to remove name from PDA now, but i think it's better to make it like in ss13, when you can overwrite PDA owner just by inserting ID in it

Dutch-VanDerLinde commented:
> Maybe it's reasonable decision to remove name from PDA now, but i think it's better to make it like in ss13, when you can overwrite PDA owner just by inserting ID in it

if you can overwrite the PDA owner by putting an ID in it then whats the point of having an owner

BombasterDS commented:
> > Maybe it's reasonable decision to remove name from PDA now, but i think it's better to make it like in ss13, when you can overwrite PDA owner just by inserting ID in it
> 
> if you can overwrite the PDA owner by putting an ID in it then whats the point of having an owner

Dunno, but in ss13 it works that way. Maybe there using in PDA chat system. They have special bomb cartridges that can be activated with PDA message.

UbaserB commented:
> > Maybe it's reasonable decision to remove name from PDA now, but i think it's better to make it like in ss13, when you can overwrite PDA owner just by inserting ID in it
> 
> if you can overwrite the PDA owner by putting an ID in it then whats the point of having an owner

Lets say you take out your ID from the pda (for example teslaloose) and throw it away, then someone who randomly picks it up can tell who owned the PDA. Would be useful for detectives and also bringing back the PDA to the right owner.

deltanedas commented:
having a button in settings to set owner to current id would be nice yes

SlamBamActionman commented:
> Lets say you take out your ID from the pda (for example teslaloose) and throw it away, then someone who randomly picks it up can tell who owned the PDA. Would be useful for detectives and also bringing back the PDA to the right owner.

I have done something similar to this when playing as Detective, but it is largely one of those things that can be solved by going ";Hey, any Chef who lost their PDA?" or just turning the PDA in to HoP.

Imo the use of a named PDA without an ID is a very niche usecase and I'd hate for people to use it to metagame chameleon PDAs.

LordEclipse commented:
Couldn't you just make Cham. PDAs apply owner name to them? This seems like a convoluted work around to address powergaming. And if you are at a point where people can access your PDA you're kind of already screwed as a traitor.

deltanedas commented:
i have personally been metagamed by a pda not having my name on it, the owner field as it is in master is just a metagaming feature. some kind of measure should be possible to change it, ideally keeping it since it does have innocent uses stated previously

SlamBamActionman commented:
> Couldn't you just make Cham. PDAs apply owner name to them? This seems like a convoluted work around to address powergaming. And if you are at a point where people can access your PDA you're kind of already screwed as a traitor.

This would make the PDA too rigid imo; what if you're a Nukie who buys the Chameleon Kit? Or a Syndie trying to disguise yourself as someone else?

Additionally the point of Chameleon items requiring wearing to be discovered is to make it more difficult to detect them at a surface inspection. For example if Sec does not suspect you have No-Slips or other Chameleon clothes they won't take it away from you after releasing you from the brig. Same case here.

> i have personally been metagamed by a pda not having my name on it, the owner field as it is in master is just a metagaming feature. some kind of measure should be possible to change it, ideally keeping it since it does have innocent uses stated previously

While I agree that the name can be used, I also think that the impact of its removal would be minor. It's right now the go-to way to identify an ID-less PDA, but if it didn't exist I doubt people would find the addition necessary since there are other ways to return a PDA to its rightful owner.

SlamBamActionman commented:
Requiring a manual change of the PDA name (even if it is just a button press) adds an unnecessary step to a disguise, especially since 99 times out of 10 your ID will be inside the PDA anyways. Unless there is a common usecase for using the name that the ID can't fulfill, I don't see much reason to keep it around.

EmoGarbage404 commented:
i think having the ID name visible on the PDA is good. It helps with identifying who it belongs to at a glance without being completely set in stone. Having it be completely fixed has obvious problems but i think that inheriting the ID name is a good solution.

SlamBamActionman commented:
> i think having the ID name visible on the PDA is good. It helps with identifying who it belongs to at a glance without being completely set in stone. Having it be completely fixed has obvious problems but i think that inheriting the ID name is a good solution.

The ID name still shows as long as the ID is inserted - I really don't think that there are many instances where a PDA is found without an ID where not having the name would cause difficulties.

1. It is the PDA of a head or single employee job. Immediately identified.
2. It is the PDA of a department. Ask over the intercom if anyone is missing it. No one responds? Hand it in to HoP.
3. It is the PDA of a passenger. If no one responds over radio then all they'll really lose is their Notes app notes since they can get a new PDA from HoP.

deltanedas commented:
in the case of some random passengers pda being lost, it should just display the last id card as the owner. no clicking needed it just automatically does it on on insertion

SlamBamActionman commented:
> in the case of some random passengers pda being lost, it should just display the last id card as the owner. no clicking needed it just automatically does it on on insertion

While I find it a bit unintuitive, I can't really argue that this suggestion doesn't solve both the metagaming and finding-pda's-owner issue. I'll look into implementing it.

BombasterDS commented:
Deleting owner's name now maybe is solution, because PDA is now slightly useless. Now PDA is just ID card holder with crew manifest inside it, when in SS13 it's useful device when you login with your id card. In my opinion, deleting owner name won't affect gameplay in general, but the addition of the ability to imprint ID information to PDA is looks better.

BombasterDS commented:
> in the case of some random passengers pda being lost, it should just display the last id card as the owner. no clicking needed it just automatically does it on on insertion

I think that automatic owner change is not that good as manual. I think there will be some situations where you need to hide your id in someone PDA, but you don't want to change it's name for some reason.

SlamBamActionman commented:
> I think that automatic owner change is not that good as manual. I think there will be some situations where you need to hide your id in someone PDA, but you don't want to change it's name for some reason.

PDAs already show the ID info (including name) separately from the PDA owner name. You can't hide your ID in another's PDA even if the owner name is different.

BombasterDS commented:
> > I think that automatic owner change is not that good as manual. I think there will be some situations where you need to hide your id in someone PDA, but you don't want to change it's name for some reason.
> 
> PDAs already show the ID info (including name) separately from the PDA owner name. You can't hide your ID in another's PDA even if the owner name is different.

I didn't mean examine text. If i understand deltanedas correctly, he want PDA to change it's owner automatically when ID card inserted in PDA. In my opinion manual owner change is more flexible in different IC sutiatuions

SlamBamActionman commented:
New behavior: The PDA updates its owner name upon a new ID being inserted. That way it still shows it if the ID is removed, but you can overwrite it with another ID.

SlamBamActionman commented:
Added a fix that keeps Agent IDs from being turned into PDAs, and removed duplicate PDAs & IDs from the chameleon menu, and removed admin PDA/ID

LordEclipse commented:
> Added a fix that keeps Agent IDs from being turned into PDAs, and removed duplicate PDAs & IDs from the chameleon menu, and removed admin PDA/ID

Admin ID is the only one with syndicate, centcom, and nukie access. Need it for access changes through UI.

SlamBamActionman commented:
> > Added a fix that keeps Agent IDs from being turned into PDAs, and removed duplicate PDAs & IDs from the chameleon menu, and removed admin PDA/ID
> 
> Admin ID is the only one with syndicate, centcom, and nukie access. Need it for access changes through UI.
> 
> Oh wait, did you mean remove it from Cham. Menu? Right, reading comprehension.

Ya sorry it was a bit unclear with how I wrote it; they're simply removed from the Chameleon menu, to prevent admin impersonation. The real PDA/ID remains as-is

a-person5660 commented:
is this still being worked on?

TheShuEd commented on Content.Client/Clothing/Systems/ChameleonClothingSystem.cs:
```diff
@@ -1,4 +1,5 @@
-﻿using System.Linq;
+using System.Linq;
```
Linq jumpscare

SlamBamActionman replied:
I didn't add this Linq! D: It was there before my PR.

It's used in this line:
`private static readonly SlotFlags[] Slots = Enum.GetValues<SlotFlags>().Except(IgnoredSlots).ToArray();`

I don't know how one would do this functionality without Linq though...

TheShuEd commented on Content.Client/PDA/PdaVisualsComponent.cs:
```diff
@@ -0,0 +1,20 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
dont use Linq

TheShuEd commented on Content.Server/PDA/PdaSystem.cs:
```diff
@@ -36,6 +37,7 @@ public sealed class PdaSystem : SharedPdaSystem
         [Dependency] private readonly UserInterfaceSystem _ui = default!;
         [Dependency] private readonly UnpoweredFlashlightSystem _unpoweredFlashlight = default!;
         [Dependency] private readonly ContainerSystem _containerSystem = default!;
+        [Dependency] private readonly IdCardSystem _idCardSystem = default!;
```
```suggestion
        [Dependency] private readonly IdCardSystem _idCard = default!;
```
modern style

TheShuEd commented on Content.Shared/Clothing/Components/ChameleonClothingComponent.cs:
```diff
@@ -32,18 +32,26 @@ public sealed partial class ChameleonClothingComponent : Component
+    /// <summary>
+    ///     Filter possible chameleon options by a tag in addition to WhitelistChameleon.
+    /// </summary>
+    [ViewVariables, DataField]
```
```suggestion
    [DataField]
```
i think useless? ViesVariables already implemented into DataField iirc

TheShuEd reviewed (commented):
Minor codestyle review

TytosB commented:
so whats stopping this from being merged, people using chameleons and not having a pda is by far the most common way to metagame chameleon outfits. we really need this

deltanedas commented:
this shouldve had breaking changes filled out for changing state to appearance

SlamBamActionman commented:
> this shouldve had breaking changes filled out for changing state to appearance

This is true; I apologize for not taking custom PDAs into account.
